From fccaf1894533b29d0500af557e8d6b1d99365235 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:00:13 -0400
Subject: [fs] btrfs: restrict snapshotting to own subvolumes

Message-id: <1407179109-13714-127-git-send-email-sandeen@redhat.com>
Patchwork-id: 87082
O-Subject: [RHEL7.1 PATCH 126/422] btrfs: restrict snapshotting to own subvolumes
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: David Sterba <dsterba@suse.cz>

commit d024206133ce21936b3d5780359afc00247655b7
Author: David Sterba <dsterba@suse.cz>
Date:   Wed Jan 15 18:15:52 2014 +0100

    btrfs: restrict snapshotting to own subvolumes

    Currently, any user can snapshot any subvolume if the path is accessible and
    thus indirectly create and keep files he does not own under his direcotries.
    This is not possible with traditional directories.

    In security context, a user can snapshot root filesystem and pin any
    potentially buggy binaries, even if the updates are applied.

    All the snapshots are visible to the administrator, so it's possible to
    verify if there are suspicious snapshots.

    Another more practical problem is that any user can pin the space used
    by eg. root and cause ENOSPC.

    Original report:
    https://bugs.launchpad.net/ubuntu/+source/apparmor/+bug/484786

    CC: stable@vger.kernel.org
    Signed-off-by: David Sterba <dsterba@suse.cz>
    Signed-off-by: Josef Bacik <jbacik@fb.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/ioctl.c b/fs/btrfs/ioctl.c
index 2d96f4b..f0f7f24 100644
--- a/fs/btrfs/ioctl.c
+++ b/fs/btrfs/ioctl.c
@@ -1574,6 +1574,12 @@ static noinline int btrfs_ioctl_snap_create_transid(struct file *file,
    btrfs_info(BTRFS_I(src_inode)->root->fs_info,
        "Snapshot src from another FS");
    ret = -EINVAL;
+  } else if (!inode_owner_or_capable(src_inode)) {
+   /*
+    * Subvolume creation is not restricted, but snapshots
+    * are limited to own subvolumes only
+    */
+   ret = -EPERM;
   } else {
    ret = btrfs_mksubvol(&file->f_path, name, namelen,
           BTRFS_I(src_inode)->root,
-- 
1.7.1